spring-cloud-skipper-server-core/src/main/java/org/springframework/cloud/skipper/server/service/ReleaseService.java

```diff
@@ -195,9 +199,58 @@ public Release install(Release release) {
 	 */
 	@Transactional
 	public Release delete(String releaseName) {
+		return this.delete(releaseName, false);
+	}
+
+	/**
+	 * If the deleteReleasePackage is true deletes the release along with the package it was created from. The
+	 * transactions succeeds only if there are no other deployed release from the same package but the releaseName.
+	 * If the deleteReleasePackage is false it behaves as {@link #delete(String)}
+	 * @param releaseName the name of the release to be deleted
+	 * @param deleteReleasePackage if true tries to delete the package of the releaseName release
+	 * @return the state of the release after requesting a deletion
+	 */
+	@Transactional
+	public Release delete(String releaseName, boolean deleteReleasePackage) {
 		Assert.notNull(releaseName, "Release name must not be null");
+		if (deleteReleasePackage) {
+			Release release = this.releaseRepository.findLatestDeployedRelease(releaseName);
+			// Note: the app deployer delete is not transactional operations (e.g. we can't revert the Deployer's state
+			// in case of failure. Therefore we should call the package delete before the release delete!
+			this.deleteReleasePackage(release);
+		}
 		Release release = this.releaseRepository.findLatestDeployedRelease(releaseName);
 		return this.releaseManager.delete(release);
+
+	}
+
+	/**
+	 * Delete all versions of the release package uploaded in the same repository as the release's package.
+	 *
+	 * @param deletedRelease Release for which the Package to be deleted
+	 */
+	private void deleteReleasePackage(Release deletedRelease) {
+
+		String pkgName = deletedRelease.getPkg().getMetadata().getName();
+		Long pkgRepoId = deletedRelease.getPkg().getMetadata().getRepositoryId();
+
+		Iterable<Release> releases = this.releaseRepository.findAll();
+		List<Release> deployedReleasesByPackageAndRepoId = StreamSupport.stream(releases.spliterator(), false)
+				.filter(r -> r.getPkg().getMetadata().getName().equalsIgnoreCase(pkgName)
+						&& r.getPkg().getMetadata().getRepositoryId() == pkgRepoId)
+				.filter(r -> r.getInfo().getStatus().getStatusCode() == StatusCode.DEPLOYED)
```

Comment:
Unfortunately we can not use the PackageMetadataService#deleteIfAllReleasesDeleted directly! 
Later deletes a Pkg only if ALL releases created from the Pkg are in inactive state. 
But here we allow a Pkg to be deleted even if the current release (on to be deleted) is still in active state! This is required to ensure the error handling because the AppDeployer operations are not transactional or revokable.  
E.g. we can NOT do this sequence: (1) delete the releaseName and then (2) attempt to delete the package. The reason is that the release deletion is not transactional and if an error occur after step (1) we can't recover from it.  The DB tables will be reverted but the AppDeployer will NOT  re-deploy the release.
The last-resource-optimization is a common technique for handling a nonTx resources in a global transactions. For this the execution sequence of the deletion should be:
(1) delete the package if all releases (excluding the current one) are in non-active state and then (2) delete the current release. In this case if something fail up until the AppDeployer undeploy the release, the transaction will leave the system in consistent state. 
Here is the test that verifies the above claims: http://bit.ly/2FDvIx4

Reply:
IMO will make sense to wait for the 447 to be merged first and the see if we can refactor this logic and reuse some of the new repo methods there.

Reply:
In the meantime i can add local-repo check to the existing code.